### PR TITLE
Sql migration

### PR DESCRIPTION
**Pull request by PapaPwn**
Purely SQL Server on the backend. Probs a lot that still needs to change, but might as well get this review on the road!

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "start": "NODE_ENV=development npx nodemon ./src/bin/www",
     "lint": "npx eslint src/**/*.js tests/**/*.js --fix",
-    "test:src": "npx nyc mocha tests/**/*.js",
+    "test:src": "npx nyc mocha tests/**/*.js --timeout 100000",
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/package.json]: Yeah for the entire test suite: https://tips.tutorialhorizon.com/2017/08/29/how-to-change-the-default-timeout-for-mocha-tests/

File: backend/package.json [hunk 2]
```diff
@@ -16,7 +16,8 @@
     "express": "~4.16.0",
     "firebase-admin": "^6.0.0",
     "morgan": "^1.9.1",
-    "sqlite3": "^4.0.2"
+    "sqlite3": "^4.0.2",
+    "tedious": "^3.0.1"
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/package.json]: Tedious is to interact with SQL Server. I'll delete sqlite3 now.

File: backend/data/faculties.js [hunk 3]
```diff
@@ -1,10 +1,10 @@
 module.exports = [
-    'Art & Design',
-    'Arts & Social Sciences',
-    'Built Environment',
-    'Business School',
-    'Engineering',
-    'Law',
-    'Medicine',
-    'Science'
+    { 'name': 'Art & Design' },
```
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/data/faculties.js]: middle ground could be to add
```
.map(name => ({ name )})
```
on the bottom of the array.

File: backend/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/package.json]: No but that's still the single test timeout. That's way too long.

File: backend/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/package.json]: yep cool.

File: backend/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/package.json]: Right true. Not sure how to workaround the DB initialisation without cutting into the timeout time?

File: backend/data/faculties.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/data/faculties.js]: Sure :)

File: backend/src/bin/www [hunk 4]
```diff
@@ -22,12 +22,13 @@ app.set('port', port)
...
 
 /**
- * Listen on provided port, on all network interfaces.
+ * Listen on provided port, on all network interfaces, once initialisation is done.
  */
-
-server.listen(port)
-server.on('error', onError)
-server.on('listening', onListening)
+app.on('ready', () => {
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/bin/www]: So it is emitted for tests as well `index.js` does it:
```
/*
 * Connect to SQL server
 */
db.init()
    .then(() => console.log('App ready!'))
    .then(() => app.emit('ready'))
```
However, when I did `app.on('ready', foo)`, in the mocha tests, it never fired.

File: backend/src/models/comment.js [hunk 5]
```diff
@@ -16,7 +16,13 @@ class Comment {
     postComment(queryObject, { body, userID }) {
         const [key, value] = Object.entries(queryObject)[0]
         return this.db
-            .insert(COMMENTS, { [key]: value, body, userID })
+            .run(`INSERT INTO ${COMMENTS} (${key}, body, userID)
+                VALUES (@${key}, @body, @userID);
+                SELECT @@identity AS id`,
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/models/comment.js]: Its the same as `?` syntax for sqlite3. Just parameters in the query for safety.

File: backend/src/models/constants.js [hunk 6]
```diff
@@ -1,3 +1,23 @@
+const { TYPES } = require('tedious')
+
+// SQL Server Config
+exports.PRODUCTION = process.env.NODE_ENV === 'production'
+exports.TESTING = process.env.NODE_ENV === 'test'
+exports.DB_NAME = exports.PRODUCTION ? 'smartcourse'
+    : exports.TESTING ? 'smartcourse-testing' : 'smartcourse-staging'
+exports.DB_CONFIG = {
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/models/constants.js]: Sure

File: backend/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/package.json]: Can you make it more targeted?

File: backend/src/bin/www (hunk 4 above)
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/src/bin/www]: `supertest.on('ready', foo)` ?

File: backend/src/models/db/test_constants.js [hunk 7]
```diff
@@ -11,44 +9,24 @@ exports.SAMPLE_QUESTIONS = [
         title: 'What year would you recommend taking this course?',
         body: 'Is this course worth taking earlier or later in my degree? What are the advantages or disadvantages?'
     },
-    {
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/models/db/test_constants.js]: Yeah forgot about that. Currently these are test only, but we can make them pinned Q's in another PR. I'll add the deleted ones back.

File: backend/src/models/likes.js [hunk 8]
```diff
@@ -12,60 +12,89 @@ class Likes {
...
+            {
+                [LIKES]: { objectType: type, objectID: id, userID }
+            })
+            .then(([{ userLiked = 0 } = {}]) => ({ userLiked }))
     }
 
     /*
      * Put likes
      */
     putLikes({ type, id, userID, value }) {
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/models/likes.js]: I definitely agree. I tried making it a single query but failed when we needed to consider the case of someone up-voting themself (I know you can do conditions in SQL but wasn't a priority getting this refactored). Can fix in another PR.

File: backend/src/bin/www (hunk 4 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/bin/www]: Sadly not

File: backend/tests/routes/api.js [hunk 9]
```diff
@@ -1,5 +1,8 @@
 const app = require('../../src')
 const supertest = require('supertest')(app)
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/tests/routes/api.js]: Doesn't work :(

File: backend/src/bin/www (hunk 4 above)
[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/src/bin/www]: :(

File: backend/tests/routes/api.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/tests/routes/api.js]: Solution found (I swear I tried what I've just done before).

File: backend/tests/routes/api.js [hunk 10]
```diff
@@ -1,5 +1,8 @@
...
+const dbInitialised = require('../../src/models/db/init_sql').initDB
```
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/tests/routes/api.js]: Solution found (I swear I tried what I've just done before).

File: backend/src/bin/www (hunk 4 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/src/bin/www]: Solution found (I swear I tried what I've just done before).

**Comment by DarkPurple141** [comment]
Just gave you a couple more conflicts to resolve.

File: backend/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/package.json]: I can indeed :)

[the remark this answers is not in the document]
Comment (PapaPwn) [reply backend/package.json]: Actually it didn't work...

[the remark this answers is not in the document]
Comment (DarkPurple141) [reply backend/package.json]: :(

File: backend/tests/routes/api.js [hunk 11]
```diff
@@ -1,8 +1,11 @@
 const app = require('../../src')
 const supertest = require('supertest')(app)
-const dbInitialised = require('../../src/models/db/init_sql').initDB
 
-before(() => dbInitialised)
+before(function (done) {
+    app.on('ready', function() {
```
Comment (DarkPurple141) [inline backend/tests/routes/api.js]: nice.

**Pull request merged**